src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,701 @@
+    { exact f_spec _ },
+    { simp only [sub_zero, cast_zero, submodule.zero_mem], }
+  end,
+  map_one' :=
+  begin
+    rw [f_congr (1 : ℤ_[p]) _ 1, cast_one],
+    { exact f_spec _ },
+    { simp only [sub_self, cast_one, submodule.zero_mem], }
+  end,
+  map_add' :=
+  begin
+    intros x y,
+    rw [f_congr (x + y) _ (f x + f y), cast_add],
+    { exact f_spec _ },
+    { convert ideal.add_mem _ (f_spec x) (f_spec y),
+      rw cast_add,
+      ring, }
+  end,
+  map_mul' :=
+  begin
+    intros x y,
+    rw [f_congr (x * y) _ (f x * f y), cast_mul],
+    { exact f_spec _ },
+    { let I : ideal ℤ_[p] := ideal.span {v},
+      have A : x * (y - f y) ∈ I := I.mul_mem_left (f_spec _),
+      have B : (x - f x) * (f y) ∈ I := I.mul_mem_right (f_spec _),
+      convert I.add_mem A B,
+      rw cast_mul,
+      ring, }
+  end, }
+
+/--
+`to_zmod` is a ring hom from `ℤ_[p]` to `zmod p`,
+with the equality `to_zmod x = (zmod_repr x : zmod p)`.
+-/
+def to_zmod : ℤ_[p] →+* zmod p :=
+to_zmod_hom p zmod_repr
+  (by { rw ←maximal_ideal_eq_span_p, exact sub_zmod_repr_mem })
+  (by { rw ←maximal_ideal_eq_span_p, exact zmod_congr_of_sub_mem_max_ideal } )
+
+/--
+`z - (to_zmod z : ℤ_[p])` is contained in the maximal ideal of `ℤ_[p]`, for every `z : ℤ_[p]`.
+
+The coercion from `zmod p` to `ℤ_[p]` is `zmod.has_coe_t`,
+which coerces `zmod p` into artibrary rings.
+This is unfortunate, but a consequence of the fact that we allow `zmod p`
+to coerce to rings of arbitrary characteristic, instead of only rings of characteristic `p`.
+This coercion is only a ring homomorphism if it coerces into a ring whose characteristic divides `p`.
+While this is not the case here we can still make use of the coercion.
+-/
+lemma to_zmod_spec (z : ℤ_[p]) : z - (to_zmod z : ℤ_[p]) ∈ maximal_ideal ℤ_[p] :=
+begin
+  convert sub_zmod_repr_mem z using 2,
+  dsimp [to_zmod, to_zmod_hom],
+  unfreezingI { rcases (exists_eq_add_of_lt (hp_prime.pos)) with ⟨p', rfl⟩ },
+  change ↑(zmod.val _) = _,
+  simp [zmod.val_cast_nat],
+  rw mod_eq_of_lt,
+  simpa only [zero_add] using zmod_repr_lt_p z,
+end
+
+lemma ker_to_zmod : (to_zmod : ℤ_[p] →+* zmod p).ker = maximal_ideal ℤ_[p] :=
+begin
+  ext x,
+  rw ring_hom.mem_ker,
+  split,
+  { intro h,
+    simpa only [h, zmod.cast_zero, sub_zero] using to_zmod_spec x, },
+  { intro h,
+    rw ← sub_zero x at h,
+    dsimp [to_zmod, to_zmod_hom],
+    convert zmod_congr_of_sub_mem_max_ideal x _ 0 _ h,
+    apply sub_zmod_repr_mem, }
+end
+
+/-- `appr n x` gives a value `v : ℕ` such that `x` and `↑v : ℤ_p` are congruent mod `p^n`.
+See `appr_spec`. -/
+-- TODO: prove that `x.appr` is a sequence that tends to `x`.
+-- Possibly relevant: summable_iff_vanishing_norm
+noncomputable def appr : ℤ_[p] → ℕ → ℕ
+| x 0     := 0
+| x (n+1) :=
+let y := x - appr x n in
+if hy : y = 0 then
+  appr x n
+else
+  let u := unit_coeff hy in
+  appr x n + p ^ n * (to_zmod ((u : ℤ_[p]) * (p ^ (y.valuation - n).nat_abs))).val
```

Comment:
In my experience, working with definitions that contain `let` is mildly annoying, and it looks like these `let`s can be inlined relatively painlessly. (If the implementation of `appr` is not supposed to be public, maybe mark it as `irreducible` after `appr_spec`?)

Reply:
`appr` shouldn't be public, no. The interface for using it is `to_zmod_pow`. I've marked it irreducible with no issues here, and I don't foresee any trouble with that in the rest of our development!